tests/test_dev_experience.py

```diff
@@ -0,0 +1,58 @@
+# Copyright: See the LICENSE file.
+
+"""Tests about developer experience: help messages, errors, etc."""
+
+import collections
+import unittest
+
+import factory
+import factory.errors
+
+Country = collections.namedtuple('Country', ['name', 'continent', 'capital_city'])
+City = collections.namedtuple('City', ['name', 'population'])
+
+
+class DeclarationTests(unittest.TestCase):
+    def test_subfactory_to_model(self):
+        """A helpful error message occurs when pointing a subfactory to a model."""
+        class CountryFactory(factory.Factory):
+            class Meta:
+                model = Country
+
+            name = factory.Faker('country')
+            continent = "Antarctica"
+
+            # Error here: pointing the SubFactory to a model, not a factory.
+            capital_city = factory.SubFactory(City)
+
+        with self.assertRaises(factory.errors.AssociatedClassError) as raised:
+            CountryFactory()
+
+        self.assertIn('City', str(raised.exception))
+        self.assertIn('Country', str(raised.exception))
+
+    def test_subfactory_to_factorylike_model(self):
+        """A helpful error message occurs when pointing a subfactory to a model.
+
+        This time with a model that looks more like a factory (ie has a `._meta`)."""
+
+        class CityModel:
+            _meta = None
+            name = "Coruscant"
+            population = 0
+
+        class CountryFactory(factory.Factory):
+            class Meta:
+                model = Country
+
+            name = factory.Faker('country')
+            continent = "Antarctica"
+
+            # Error here: pointing the SubFactory to a model, not a factory.
+            capital_city = factory.SubFactory(CityModel)
+
+        with self.assertRaises(factory.errors.AssociatedClassError) as raised:
+            CountryFactory()
+
+        self.assertIn('CityModel', str(raised.exception))
+        self.assertIn('Country', str(raised.exception))
```

Comment:
Perhaps the entire exception message could be hardcoded. It helps finding the code that generates the error as well as verifying what users see.

Here’s what I’m suggesting (moving the factory declaration to the module to get nicer qualified names):
```diff
diff --git a/tests/test_dev_experience.py b/tests/test_dev_experience.py
index 6677b08..b8e3bc3 100644
--- a/tests/test_dev_experience.py
+++ b/tests/test_dev_experience.py
@@ -12,47 +12,61 @@ Country = collections.namedtuple('Country', ['name', 'continent', 'capital_city'
 City = collections.namedtuple('City', ['name', 'population'])
 
 
-class DeclarationTests(unittest.TestCase):
-    def test_subfactory_to_model(self):
-        """A helpful error message occurs when pointing a subfactory to a model."""
-        class CountryFactory(factory.Factory):
-            class Meta:
-                model = Country
+class CountryFactory(factory.Factory):
+    class Meta:
+        model = Country
 
-            name = factory.Faker('country')
-            continent = "Antarctica"
+    name = factory.Faker('country')
+    continent = "Antarctica"
 
-            # Error here: pointing the SubFactory to a model, not a factory.
-            capital_city = factory.SubFactory(City)
+    # Error here: pointing the SubFactory to a model, not a factory.
+    capital_city = factory.SubFactory(City)
 
-        with self.assertRaises(factory.errors.AssociatedClassError) as raised:
-            CountryFactory()
 
-        self.assertIn('City', str(raised.exception))
-        self.assertIn('Country', str(raised.exception))
+class CityModel:
+    _meta = None
+    name = "Coruscant"
+    population = 0
 
-    def test_subfactory_to_factorylike_model(self):
-        """A helpful error message occurs when pointing a subfactory to a model.
 
-        This time with a model that looks more like a factory (ie has a `._meta`)."""
+class CountryModelFactory(factory.Factory):
+    class Meta:
+        model = Country
 
-        class CityModel:
-            _meta = None
-            name = "Coruscant"
-            population = 0
+    name = factory.Faker('country')
+    continent = "Antarctica"
 
-        class CountryFactory(factory.Factory):
-            class Meta:
-                model = Country
+    # Error here: pointing the SubFactory to a model, not a factory.
+    capital_city = factory.SubFactory(CityModel)
 
-            name = factory.Faker('country')
-            continent = "Antarctica"
 
-            # Error here: pointing the SubFactory to a model, not a factory.
-            capital_city = factory.SubFactory(CityModel)
+class DeclarationTests(unittest.TestCase):
+    def test_subfactory_to_model(self):
+        """A helpful error message occurs when pointing a subfactory to a model."""
 
         with self.assertRaises(factory.errors.AssociatedClassError) as raised:
             CountryFactory()
 
-        self.assertIn('CityModel', str(raised.exception))
-        self.assertIn('Country', str(raised.exception))
+        self.assertEqual(
+            str(raised.exception),
+            "<BuildStep for <StepBuilder("
+            "<FactoryOptions for CountryFactory>, strategy='create')>"
+            ">: Attempting to recursing into a non-factory object "
+            "<class 'tests.test_dev_experience.City'>"
+        )
+
+    def test_subfactory_to_factorylike_model(self):
+        """A helpful error message occurs when pointing a subfactory to a model.
+
+        This time with a model that looks more like a factory (ie has a `._meta`)."""
+
+        with self.assertRaises(factory.errors.AssociatedClassError) as raised:
+            CountryModelFactory()
+
+        self.assertEqual(
+            str(raised.exception),
+            "<BuildStep for <StepBuilder("
+            "<FactoryOptions for CountryModelFactory>, strategy='create')>"
+            ">: Attempting to recursing into a non-factory object "
+            "<class 'tests.test_dev_experience.CityModel'>"
+        )
```

Reply:
OK for adding the whole line (or maybe some regexp, for stability).
KO for moving the factory classes: they are related to the specific test, not to the module; each test function is likely to declare its own factories.